noxfile.py

```diff
@@ -1,9 +1,47 @@
+from pathlib import Path
+
 import nox
 
-@nox.session(python=["3.11"]) # The python version should match the readthedocs configuration.
+requirements_files = list(
+    {path.stem for path in Path.cwd().glob("**/*.in")}
+)
+
+
+@nox.session(name="pip-compile", python=["3.11"])
+@nox.parametrize(["req"], arg_values_list=requirements_files, ids=requirements_files)
+def pip_compile(session: nox.Session, req: str):
+    """Generate lock files from input files or upgrade packages in lock files."""
+    # fmt: off
+    session.install(
+      "-r", "requirements/pip-tools.in",
+      "-c", "requirements/pip-tools.txt",
+    )
+    # fmt: on
+
+    # Use --upgrade by default unless a user passes -P.
+    upgrade_related_cli_flags = "-P", "--upgrade-package", "--no-upgrade"
+    has_upgrade_related_cli_flags = any(
+        arg.startswith(upgrade_related_cli_flags) for arg in session.posargs
+    )
+    injected_extra_cli_args = () if has_upgrade_related_cli_flags else ("--upgrade",)
+
+    session.run(
```

Comment:
Does `session.run()` guarantee a specific working directory for invoking other programs? If not, perhaps it'd help with predictable execution to set the CWD to `requirements_directory.parent`.

Reply:
Yes, it always defaults to the directory in which the noxfile.py resides (that's the repository root in this case).